### ISSUE TITLE
Generated code has syntax errors

### ISSUE THREAD

**Issue by john-aws** [mtth/avsc#291]
I am seeing the following error when parsing an Avro file:

```
SyntaxError: Unexpected token 'case'
    at new Function (<anonymous>)
    at RecordType._createConstructor (node_modules/avsc/lib/types.js:2184:16)
    at new RecordType (node_modules/avsc/lib/types.js:2116:33)
    at node_modules/avsc/lib/types.js:226:14
    at Function.Type.forSchema (node_modules/avsc/lib/types.js:227:7)
    at node_modules/avsc/lib/types.js:213:19
    at Array.map (<anonymous>)
    at Function.Type.forSchema (node_modules/avsc/lib/types.js:212:24)
    at new Field (node_modules/avsc/lib/types.js:2787:20)
    at RecordType.<anonymous> (node_modules/avsc/lib/types.js:2110:17)
Emitted 'error' event on BlockDecoder instance at:
    at errorOrDestroy (internal/streams/destroy.js:108:12)
    at BlockDecoder.onerror (_stream_readable.js:746:7)
    at BlockDecoder.emit (events.js:311:20)
    at BlockDecoder._decodeHeader (node_modules/avsc/lib/containers.js:198:10)
    at BlockDecoder._write (node_modules/avsc/lib/containers.js:219:13)
    at doWrite (_stream_writable.js:441:12)
    at writeOrBuffer (_stream_writable.js:425:5)
    at BlockDecoder.Writable.write (_stream_writable.js:316:11)
    at ReadStream.ondata (_stream_readable.js:714:22)
    at ReadStream.emit (events.js:311:20)
```

The root cause appears to be that the avsc project dynamically generates JavaScript code as a string in `RecordType.prototype._createConstructor` (in lib/types.js) and this code will have syntax errors for certain types of data.

In my situation, the Avro file contains a construct (a data element?) named 'case' which the avsc code then uses to construct a JavaScript string containing a constructor in `outerBody` as follows:

```
return function case(v0,v1,v2,v3,v4,v5,v6,v7) {
  if (v0 === undefined) { this.name = d0(); } else { this.name = v0; }
  if (v1 === undefined) { this.description = d1(); } else { this.description = v1; }
  if (v2 === undefined) { this.awards = d2(); } else { this.awards = v2; }
  if (v3 === undefined) { this.zip = d3(); } else { this.zip = v3; }
  if (v4 === undefined) { this.location = d4(); } else { this.location = v4; }
  if (v5 === undefined) { this.project = d5(); } else { this.project = v5; }
  if (v6 === undefined) { this.created_datetime = d6(); } else { this.created_datetime = v6; }
  if (v7 === undefined) { this.updated_datetime = d7(); } else { this.updated_datetime = v7; }
```

The problem is that the generated code uses the Avro data element's name as the name of a constructor function, so the JavaScript keyword 'case' is being used for a function name, and that's a syntax error. Other JavaScript keywords will fail similarly, so the fix is not to modify the Avro data to not use the term 'case' (which is a perfectly acceptable term, especially in a clinical context, for example). Perhaps the fix is to modify the JS-generation code to include a standard prefix in all these constructors, to prevent keyword clashes, or use a Map from data element name to constructor function.

This Avro file is, in all respects, valid afaik.

I'm also concerned at the mechanism used here from a security perspective. Are there opportunities here for deliberately malformed data in an Avro file to end up being executed in the context of some customer's server? An Avro variant of SQL injection.

**Comment by mtth** [comment]
Hi @john-aws, thanks for reporting this. The record name is indeed causing the issue here. A good way forward is probably to suffix these reserved words, keeping the current naming stable. I'll get a fix out soon.

If you control the schema, I would also recommend renaming the record to `Case`. Records are similar to classes (where PascalCase is more idiomatic) and it would solve this particular issue (all JavaScript reserved keywords are lowercase).

> I'm also concerned at the mechanism used here from a security perspective. Are there opportunities here for deliberately malformed data in an Avro file to end up being executed in the context of some customer's server? An Avro variant of SQL injection.

I won't claim to be a security expert but code generation only accepts [alpha-numeric identifiers plus underscore](https://github.com/mtth/avsc/blob/master/lib/types.js#L46), so I believe should be safe.

**Comment by mtth** [comment]
Fixed in `5.4.20`.

**Comment by john-aws** [comment]
Confirmed, thank you for the quick fix.
